ccm-connector/src/main/java/com/sequenceiq/cloudbreak/ccmimpl/endpoint/RetryingServiceEndpointFinder.java

```diff
@@ -0,0 +1,129 @@
+package com.sequenceiq.cloudbreak.ccmimpl.endpoint;
+
+import java.time.ZonedDateTime;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nonnull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.rholder.retry.Attempt;
+import com.github.rholder.retry.RetryException;
+import com.github.rholder.retry.Retryer;
+import com.github.rholder.retry.RetryerBuilder;
+import com.github.rholder.retry.StopStrategies;
+import com.github.rholder.retry.StopStrategy;
+import com.github.rholder.retry.WaitStrategies;
+import com.github.rholder.retry.WaitStrategy;
+import com.google.common.base.Throwables;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpoint;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointFinder;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointLookupException;
+import com.sequenceiq.cloudbreak.ccm.endpoint.ServiceEndpointRequest;
+
+/**
+ * Service endpoint finders that retries lookups on a delegate finder.
+ */
+public class RetryingServiceEndpointFinder implements ServiceEndpointFinder {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RetryingServiceEndpointFinder.class);
+
+    /**
+     * The delegate service endpoint finder.
+     */
+    private final ServiceEndpointFinder delegate;
+
+    /**
+     * Creates a retrying service endpoint finder with the specified parameters.
+     *
+     * @param delegate the delegate service endpoint finder
+     */
+    public RetryingServiceEndpointFinder(@Nonnull ServiceEndpointFinder delegate) {
+        this.delegate = Objects.requireNonNull(delegate, "delegate is null");
+    }
+
+    @Nonnull
+    @Override
+    public <T extends ServiceEndpoint> T getServiceEndpoint(ServiceEndpointRequest<T> serviceEndpointRequest)
+            throws ServiceEndpointLookupException, InterruptedException {
+
+        String targetInstanceId = serviceEndpointRequest.getTargetInstance().getTargetInstanceId();
+
+        Retryer<T> loop = buildRetryer(serviceEndpointRequest);
+
+        T serviceEndpoint;
+        try {
+            serviceEndpoint = loop.call(() -> delegate.getServiceEndpoint(serviceEndpointRequest));
+        } catch (ExecutionException e) {
+            try {
```

Comment:
This whole try catch hell looks very awful for me. My feeling is the guava-retrying library is not a good choice.

Reply:
Maybe you'll prefer this style.